lib/services/access_request.go

```diff
@@ -1583,6 +1650,65 @@ func (m *RequestValidator) setRolesForResourceRequest(ctx context.Context, req t
 	return nil
 }
 
+// pruneRequestedRolesNotMatchingKubernetesResourceKinds will filter out the kubernetes kinds from the requested resource IDs (kube_cluster and its subresources)
+// disregarding whether it's leaf or root cluster request, and for each requested role, ensures that all requested kube resource kind are allowed by the role.
+// Roles not matching with every kind requested, will be pruned from the requested roles.
+//
+// Returns pruned roles, and a map of requested roles with allowed kinds (with denied applied), used to help aid user in case a request gets rejected,
+// lets user know which kinds are allowed for each requested roles.
+func (m *RequestValidator) pruneRequestedRolesNotMatchingKubernetesResourceKinds(requestedResourceIDs []types.ResourceID, requestedRoles []string) ([]string, map[string][]string) {
+	// Filter for the kube_cluster and its subresource kinds.
+	requestedKubeKinds := make(map[string]struct{})
+	for _, resourceID := range requestedResourceIDs {
+		if resourceID.Kind == types.KindKubernetesCluster || slices.Contains(types.KubernetesResourcesKinds, resourceID.Kind) {
+			requestedKubeKinds[resourceID.Kind] = struct{}{}
+		}
+	}
+
+	if len(requestedKubeKinds) == 0 {
+		return requestedRoles, nil
+	}
+
+	goodRoles := make(map[string]struct{})
+	mappedRequestedRolesToAllowedKinds := make(map[string][]string)
+	for _, requestedRoleName := range requestedRoles {
+		allowedKinds, deniedKinds := getKubeResourceKinds(m.kubernetesResource.allow[requestedRoleName]), getKubeResourceKinds(m.kubernetesResource.deny)
+
+		// Any resource is allowed.
+		if allowedKinds == nil && deniedKinds == nil {
```

Comment:
sure i guess checking len is slightly safer b/c i'm thinking you're concerned this can happen

var array []string, by itself is nil
array := []string{} is not?

Reply:
Yeah. If someone changes the approach from using a non-initialized array to an initialized array, the code won't catch it.